flink/src/main/java/org/apache/iceberg/flink/source/RowDataReader.java

```diff
@@ -0,0 +1,109 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink.source;
+
+import java.util.Map;
+import java.util.Set;
+import org.apache.flink.table.data.RowData;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.avro.Avro;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.flink.data.FlinkAvroReader;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.CloseableIterator;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.mapping.NameMappingParser;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.types.TypeUtil;
+import org.apache.iceberg.util.PartitionUtil;
+
+class RowDataReader extends BaseDataReader<RowData> {
+
+  private final int[] finalProjection;
+  private final Schema expectedSchema;
+  private final String nameMapping;
+
+  RowDataReader(CombinedScanTask task, FileIO fileIo, EncryptionManager encryption, int[] finalProjection,
+                Schema expectedSchema, String nameMapping) {
+    super(task, fileIo, encryption);
+    this.finalProjection = finalProjection;
+    this.expectedSchema = expectedSchema;
+    this.nameMapping = nameMapping;
+  }
+
+  @Override
+  protected CloseableIterator<RowData> nextTaskIterator(FileScanTask task) {
+    // schema or rows returned by readers
+    PartitionSpec spec = task.spec();
+    Set<Integer> idColumns = spec.identitySourceIds();
+    Schema partitionSchema = TypeUtil.select(expectedSchema, idColumns);
+
+    Map<Integer, ?> idToConstant = partitionSchema.columns().isEmpty() ? ImmutableMap.of() :
+        PartitionUtil.constantsMap(task, RowDataReader::convertConstant);
+    CloseableIterable<RowData> iterable = newIterable(task, idToConstant);
+    ProjectionRowData projectionRow = new ProjectionRowData();
+    return (finalProjection == null ? iterable : CloseableIterable.transform(
+        iterable, rowData -> (RowData) projectionRow.replace(rowData, finalProjection))).iterator();
```

Comment:
Q:   Why we need to transform the `CloseableIterable<RowData>`  to be a projected `RowData` iterable again,  I mean we've created an AVRO iterable with the projected read schema, it should guarantee  that iterable only contains the projected columns ?

Reply:
You can take a look to `FlinkSchemaUtil. pruneWithoutReordering`, it just keep the order from original schema. But the real Flink projection may change the order.